### PR TITLE
[JENKINS-55361] DescriptorExtensionList not locking correctly, leading to deadlocks

### PR DESCRIPTION
**Pull request by Greybird**
See [JENKINS-55361](https://issues.jenkins-ci.org/browse/JENKINS-55361).

This PR aims to remove an intermittent deadlock at startup.

Not exactly sure of how to test this appropriately, so did not include any related test.

(Unsure, but might be linked in some way to the underlying problem #3777 is aiming to solve)

### Proposed changelog entries

* Removed a potential startup deadlock situation in extension loading

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [X] Appropriate autotests or explanation to why this change has no tests
- ~~For dependency updates: links to external changelogs and, if possible, full diffs~~

### Desired reviewers

@daniel-beck, @oleg-nenashev

### CODE REVIEW

**Review by jglick — APPROVED** [review]
Cannot claim to understand the problem or the patch. Would be nice to have a second review for something this risky.

**Review by batmat — APPROVED** [review]
LGTM.

So I expect you are running this already in your production system(s)?

**Review by jvz — APPROVED** [review]
Makes sense to me. Not a big fan of exposing lock objects, but I see that's already there.

**Comment by Greybird** [comment]
@batmat : I'm not running this in production right now, but you are right, this would provide some additional input regarding the potential impact.
As I have no access to my Jenkins production instance right now, I will arrange with people in charge, and come back to you after a few days of use.

**Review by oleg-nenashev — APPROVED** [review]
Looks good AFAICT

**Comment by oleg-nenashev** [comment]
Thanks @Greybird !

**Pull request merged**
